### ISSUE TITLE
Different mask values when same model is loaded again

### ISSUE THREAD

**Issue by sandeep-krutrim** [nomic-ai/contrastors#30]
I trained a nomic-bert from scratch. When I am loading the model for mask filling inference, the model return different values for mask each time the model is loaded. 

The code to do mask filling is - 
`from transformers import AutoModelForMaskedLM, AutoConfig, AutoTokenizer, pipeline
from model_latest.modeling_hf_nomic_bert import NomicBertForPreTraining

tokenizer = AutoTokenizer.from_pretrained('some-tokenizer') # 

model_path = 'custom-model'

config = AutoConfig.from_pretrained(model_path, trust_remote_code=True) # the config needs to be passed in
model = NomicBertForPreTraining.from_pretrained(model_path,config=config, trust_remote_code=True)

# To use this model directly for masked language modeling
classifier = pipeline('fill-mask', model=model, tokenizer=tokenizer,device="cpu")

print(classifier("I [mask] playing football."))`

Is the model loading causing some random initialization? Any fix for this?

**Comment by zanussbaum** [comment]
I'm not super familiar with the pipeline code but a few things I'd check if the model in eval mode/dropout turned off when using the pipeline. Otherwise can dive into what's going on. It shouldn't be using any random init for weights but you can double check this by loading two models and seeing if they have the same weights

**Comment by sandeep-krutrim** [comment]
even with custom inference code, the same behaviour is seen - 

`def mask_filling_inference(sentence, model_path=None, top_k=5):

    config = AutoConfig.from_pretrained(model_path, trust_remote_code=True) # the config needs to be passed in
    model = AutoModelForMaskedLM.from_pretrained(model_path,config=config,trust_remote_code=True)

    # Tokenize input
    tokenized_input = tokenizer.tokenize(sentence)
    masked_index = tokenized_input.index('<mask>')  # assuming only one [MASK] token

    # Convert tokens to tensor
    indexed_tokens = tokenizer.convert_tokens_to_ids(tokenized_input)
    tokens_tensor = torch.tensor([indexed_tokens])

     # Create attention mask
    attention_mask = torch.ones_like(tokens_tensor)
    attention_mask[tokens_tensor == tokenizer.mask_token_id] = 0

    # Predict
    with torch.no_grad():
        outputs = model(tokens_tensor, attention_mask=attention_mask)
        predictions = outputs[0]

    # Get top predictions for the masked token
    _, predicted_indexes = torch.topk(predictions[0, masked_index], top_k)
    predicted_tokens = tokenizer.convert_ids_to_tokens(predicted_indexes.tolist())

    return predicted_tokens

# Example usage
sentence = "I want to <mask> a car."
predicted_tokens = mask_filling_inference(sentence,model_path)
print("Top predictions for the masked token:", predicted_tokens)`

**Comment by zanussbaum** [comment]
I think you need to set `model.eval` otherwise any non-deterministic layers (like dropout) will cause different answers
```python
import torch
from transformers import AutoConfig, AutoModelForMaskedLM, AutoTokenizer

model_path = "nomic-ai/nomic-bert-2048"
config = AutoConfig.from_pretrained(model_path, trust_remote_code=True) # the config needs to be passed in
model = AutoModelForMaskedLM.from_pretrained(model_path,config=config,trust_remote_code=True)
model.eval()

tokenizer = AutoTokenizer.from_pretrained(model_path, trust_remote_code=True)

sentence = "quick brown fox jumps over the lazy [MASK]"

# Tokenize input
tokenized_input = tokenizer.tokenize(sentence)
masked_index = tokenized_input.index('[MASK]')  # assuming only one [MASK] token

# Convert tokens to tensor
indexed_tokens = tokenizer.convert_tokens_to_ids(tokenized_input)
tokens_tensor = torch.tensor([indexed_tokens])

 # Create attention mask
attention_mask = torch.ones_like(tokens_tensor)
attention_mask[tokens_tensor == tokenizer.mask_token_id] = 0

# Predict
with torch.no_grad():
    outputs = model(tokens_tensor, attention_mask=attention_mask)
    predictions = outputs[0]

# Get top predictions for the masked token
_, predicted_indexes = torch.topk(predictions[0, masked_index], 5)
predicted_tokens = tokenizer.convert_ids_to_tokens(predicted_indexes.tolist())
print(predicted_tokens)
```
